store.go

```diff
@@ -0,0 +1,221 @@
+		keys[i] = store.clean(keys[i])
+	}
+	return
+}
+
+// Walk traverses the keyspace under the given prefix, calling the walk function
+// for each key.
+//
+// If the walk function returns an error the iteration is stopped and the error
+// is returned by Walk.
+func (store *Store) Walk(ctx context.Context, prefix string, walk func(key string) error) (err error) {
+	var result io.ReadCloser
+	var query = Query{
+		{Name: "keys"},
+		{Name: "recurse", Value: "true"},
+	}
+
+	if _, result, err = store.client().do(ctx, "GET", store.path(prefix), query, nil); err != nil {
+		return
+	}
+	defer result.Close()
+
+	var stream = json.NewStreamDecoder(result)
+	var key string
+
+	for stream.Decode(&key) == nil {
+		if err = walk(store.clean(key)); err != nil {
+			return
+		}
+	}
+
+	err = stream.Err()
+	return
+}
+
+// Read reads the value of the given key, returning it as a pair of an
+// io.ReadCloser and value of the last index that modified the key.
+//
+// The program must close the value when it's done reading from it to prevent
+// any leak of internal resources.
+func (store *Store) Read(ctx context.Context, key string) (value io.ReadCloser, index int64, err error) {
+	var header http.Header
+	var sindex string
+
+	if header, value, err = store.client().do(ctx, "GET", store.path(key), Query{{Name: "raw"}}, nil); err != nil {
+		return
+	}
+
+	if sindex = header.Get("X-Consul-Index"); len(sindex) == 0 {
+		value.Close()
+		err = errMissingXConsulIndex
+		return
+	}
+
+	if index, err = strconv.ParseInt(sindex, 10, 64); err != nil {
+		value.Close()
+		err = fmt.Errorf("bad X-Consul-Index in HTTP response: %s", sindex)
+		return
+	}
+
+	return
+}
+
+// ReadValue reads the JSON-encoded value at the given key into ptr. The usual
+// unmarshaling rules apply.
+//
+// See Read for more details on the method.
+func (store *Store) ReadValue(ctx context.Context, key string, ptr interface{}) (index int64, err error) {
+	var result io.ReadCloser
+
+	if result, index, err = store.Read(ctx, key); err != nil {
+		return
+	}
+	defer result.Close()
+
+	err = json.NewDecoder(result).Decode(ptr)
+	return
+}
+
+// Write writes bytes from value at the given key in the consul key/value store.
+//
+// If index is set to a positive value, it is used to turn the write call into a
+// compare-and-swap operation. The value will only be updated if the last index
+// that modified the key matches the given index.
+//
+// If the given context inherits from a lock, the associated session is used to
+// allow the write operation to succeed.
+//
+// The method returns a boolean that indicates whether the write operation
+// succeeded, it would return false if the key was locked by another session
+// or if index was specified but was different in consul.
+//
+// The value is always closed by a call to Write, even if the method returns an
+// error.
+func (store *Store) Write(ctx context.Context, key string, value io.ReadCloser, index int64) (ok bool, err error) {
```

Comment:
Maybe, what about the 1% time where you need the function to close your reader :P... this also get passed directly as the body to the HTTP request which expects an io.ReadCloser, so I think it makes it more flexible to not hide this implementation detail (and also it's consistent with the Read method).

I think the common case will be to use ReadValue/WriteValue, so I don't think it's too inconvenient to have those lower-level methods use io.ReadCloser, and it covers the case where you do need to close your resource when the body has been fully sent.